intranet/apps/users/models.py

```diff
@@ -1554,6 +1578,68 @@ def clear_cache(self):
         for attr in User.ldap_user_attributes:
             cache.delete(":".join((self.dn, attr)))
             cache.delete(User.create_secure_cache_key(":".join((self.dn, attr))))
+        logger.debug("Clearing cached DB values")
+        if self.cache:
+            UserCache.objects.get(pk=self.cache.pk).delete()
+        self.set_cache()
+
+    def set_cache(self):
+        logger.debug("Setting DB cache using LDAP values")
+        if not self.cache:
+            self.cache = UserCache.objects.create(objectClass=self.user_type, first_name=self.first_name, last_name=self.last_name)
+        self.cache.grade_number = self.grade.number if self.grade else None
+        self.cache.graduation_year = int(self.graduation_year) if self.graduation_year else None
+        bool_gender = None
+        if self.sex:
+            bool_gender = True if self.sex.lower()[:1] == "m" else False
+        self.cache.gender = bool_gender
+        self.cache.save()
+        self.save()
+
+    def get_or_set_cache(self, attribute):
+        mappings = {
+            "objectClass": "user_type",
+            "gender": "sex",
+            "first_name": "first_name",
+            "last_name": "last_name",
+            "graduation_year": "graduation_year"
+        }
+        try:
+            if not self.cache:
+                logger.debug("Initializing UserCache for {}".format(self))
+                self.cache = UserCache.objects.create()
+                self.cache.save()
+                self.save()
+            val = getattr(self.cache, attribute)
+            if val:
+                return val
+            logger.debug("Value not found in cache, attempting to set from LDAP")
+            if attribute in mappings:
+                if attribute == "gender":
+                    logger.debug("Setting gender from LDAP")
+                    if self.sex:
+                        self.cache.gender = True if self.sex.lower()[:1] == "m" else False
+                    else:
+                        return None
+                else:
+                    logger.debug("Setting {} from LDAP".format(attribute))
+                    if str(self.__getattr__(mappings[attribute])).isdigit():
+                        setattr(self.cache, attribute, int(self.__getattr__(mappings[attribute])))
+                    else:
+                        setattr(self.cache, attribute, self.__getattr__(mappings[attribute]))
+                self.cache.save()
+                return self.get_or_set_cache(attribute)
+            elif attribute == "grade_number":
+                logger.debug("Setting grade_number from LDAp")
+                self.cache.grade_number = self.grade.number if self.grade else None
+                self.cache.save()
+                return self.get_or_set_cache(attribute)
+            return None
+        except UserCache.DoesNotExist:
```

Comment:
What places? The only one i'm seeing that it should be expected is 1610, which could also be refactored to getattr + if, though that's not a big deal